python/podio/pythonizations/collection_subscript.py

```diff
@@ -0,0 +1,23 @@
+"""Pythonize subscript operation for collections"""
+
+import cppyy
+from .utils.pythonizer import Pythonizer
+
+
+class CollectionSubscriptPythonizer(Pythonizer):
+    """Bound-check __getitem__ for classes derived from podio::CollectionBase"""
+
+    @classmethod
+    def priority(cls):
+        return 50
+
+    @classmethod
+    def callback(cls, class_, name):
+        def get_item(self, i):
+            try:
+                return self.at(i)
+            except cppyy.gbl.std.out_of_range as exc:
+                raise IndexError("collection index out of range") from exc
```

Comment:
How decipherable will the stacktrace be in this case? I.e. how many levels do we add with this callback mechanism? Or is cppyy clever enough to hide a bunch of it's internals and make it quite obvious where the actual error is? I am asking because people usually don't read the full stacktrace, especially if it's long.

Reply:
Example:
```python
>>> import edm4hep
>>> coll = edm4hep.MCParticleCollection()
>>> def foo(x):
...      x[0]
... 
>>> foo(coll)
Traceback (most recent call last):
  File "/scratch/podio/python/podio/pythonizations/collection_subscript.py", line 18, in get_item
    return self.at(i)
cppyy.gbl.std.out_of_range: none of the 2 overloaded methods succeeded. Full details:
  edm4hep::MCParticle edm4hep::MCParticleCollection::at(unsigned long index) =>
    out_of_range: deque::_M_range_check: __n (which is 0)>= this->size() (which is 0)
  edm4hep::MutableMCParticle edm4hep::MCParticleCollection::at(unsigned long index) =>
    out_of_range: deque::_M_range_check: __n (which is 0)>= this->size() (which is 0)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in foo
  File "/scratch/podio/python/podio/pythonizations/collection_subscript.py", line 20, in get_item
    raise IndexError("collection index out of range") from exc
IndexError: collection index out of range
>>> 
```
The context printout could be suppressed by raising `from None` if you find the following more readable:
```python
>>> foo(coll)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in foo
  File "/scratch/podio/python/podio/pythonizations/collection_subscript.py", line 20, in get_item
    raise IndexError("collection index out of range") from None
IndexError: collection index out of range
```

In both cases I'd say it's rather easy to spot the culprit in `line 2, in foo`

Reply:
I have a slight preference for the `from None` version, simply because it is less to read for essentially the same information. The other one just dumps a bunch of implementation details as well, IMHO.